protected/humhub/services/MigrationService.php

```diff
@@ -0,0 +1,328 @@
+<?php
+
+/*
+ * @link      https://www.humhub.org/
+ * @copyright Copyright (c) 2023 HumHub GmbH & Co. KG
+ * @license   https://www.humhub.com/licences
+ */
+
+namespace humhub\services;
+
+use humhub\commands\MigrateController;
+use humhub\components\Application;
+use humhub\components\Event;
+use humhub\components\Module;
+use humhub\events\MigrationEvent;
+use humhub\interfaces\ApplicationInterface;
+use Throwable;
+use Yii;
+use yii\base\ActionEvent;
+use yii\base\Component;
+use yii\base\Controller;
+use yii\base\InvalidConfigException;
+use yii\base\Module as BaseModule;
+use yii\console\ExitCode;
+
+/**
+ * @since 1.16
+ */
+class MigrationService extends Component
+{
+    public const EVENT_AFTER_MIGRATION = 'afterMigration';
+    public const EVENT_BEFORE_MIGRATION = 'beforeMigration';
+    protected const MIGRATION_NEW = 'new';
+    protected const MIGRATION_UNINSTALL = 'uninstall';
+    protected const MIGRATION_UP = 'up';
+
+    protected BaseModule $module;
+    protected ?string $path = null;
+    private ?int $lastMigrationResult = null;
+    private ?string $lastMigrationOutput = null;
+
+    /**
+     * @param Module|ApplicationInterface|Application|null $module
+     * @param array $config
+     */
+    public function __construct(?BaseModule $module = null, array $config = [])
+    {
+        $this->module = $module ?? Yii::$app;
+        parent::__construct($config);
+    }
+
+    public function init()
+    {
+        parent::init();
+
+        if ($this->path === null) {
+            $moduleId = $this->module instanceof Module
+                ? $this->module->id
+                : 'humhub';
+
+            $this->path = "@$moduleId/migrations";
+        }
+
+        $realpath = $this->getPath(true);
+
+        if ($realpath === false || !is_dir($realpath)) {
+            Yii::debug("Module has no migrations directory.", $this->module->id);
+            $this->path = null;
+        }
+    }
+
+    /**
+     * @return Module|ApplicationInterface|Application|null
+     */
+    public function getModule(): BaseModule
+    {
+        return $this->module;
+    }
+
+    public function getPath(bool $resolve = false): ?string
+    {
+        if (!$resolve || $this->path === null) {
+            return $this->path;
+        }
+
+        $path = realpath(Yii::getAlias($this->path));
+
+        if ($path === false) {
+            return null;
+        }
+
+        return $path;
+    }
+
+    public function getLastMigrationOutput(): ?string
+    {
+        return $this->lastMigrationOutput;
+    }
+
+    public function getLastMigrationResult(): ?int
+    {
+        return $this->lastMigrationResult;
+    }
+
+    public function isValid(): bool
```

Comment:
What is this needed for?

Reply:
The idea was, that when getting the `MigrationService` from `$module->getMigrationService`, I can check if there is a valid migration path (path `/module/migrations` exist). Hence, I can decide whether or not to run the migration.

Reply:
Ok got it, Maybe we could mark it as `private`?

Reply:
Well, if it's private, it makes no sense anymore. Because e.g. from the Module class or the MigrateController or InstallController I could no longer check if there is a valid path (since the `getPath` is now also private).

Internally, I only use it in one single place, so a private method would be overkill.

So it boils down to the question, if we want to be discoverable, if there is a valid path.

Alternatively, we could create a method `pendingMigrations(): ?bool` which would return
- `true` if path is valid and pending migrations where found
- `false` if path is valid but no pending migrations found
- `null` if path is invalid

Reply:
What about two methods?

- hasMigrations(): bool     /* or `isValid();` migrations available */  
- hasPendingMigrations(): bool

Reply:
Done: [Add hasMigrations(Pending)()](https://github.com/humhub/humhub/pull/6550/commits/bfe92a47fa159d5cc7d95c707074217085ba4ba1);